### PR TITLE
Avoid extra offset in arrow function body when using long types (alternative solution)

### LINKED ISSUE #10848: The offset for React component body depends on component name length in v2.3
**Prettier 2.3.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLWcBOAzAhmOAAgFkBPAYQgFsAHaBGABUwhoGdDgAdKQwgDySEoAVyoAjLDwC+PHpChsYJCtTpQGQgEpx8MAHQA1CKgAmAMRFQwMVNEq16sADxkH6hs1ZsAfIQC8hAAUwAKE0gCUAX7cvIQKSoS4AYQibHAAgkERPHyYcDAimLzOpqgAbj78KcD80gDcSTW40s4A9GWVMnJQ6PA4+ERuak4wAExe7Jy5AkKiElJxpHNikpjdUPLQicOOGrBj2ro2RiYWVjZ2UO6jrqp7DBMs7H6BIQIANISk4VH+MTMEspkoE0plsjN8oVioRShUqjU6o0foFgKQGk1US12p0fBseH0sHgCCobvsYAB1VAwAAWABloABzAByuCocEmHFifEEwlWiz4yz5C3WUFkm2s22Uuw8sCptIZUBZbLgRz0pzMlmstnsI3JdzJDHl9KZrPZnNewVC-C+P0i0WmcSBmNS6SyOTiUKKJVx1VRSO+NXRjRBnGxHXhMhAHxArB1imQoFwmBYAHdGMmEGxkCBcAAbVO4UjZmPiTD4ADWBQAyjR8OhGcgYJgRHAY3AFqZTHBTHTcEqRLhGXBzBBMFRcDBbEqc7gRDAINGQDSYFQ8xSadS4Gw6wRq1nqRVqaQc2A2CW0IosExy4yJ8g8Hn0jGAFZsfgAIXLYCrMGrKrpdA4AffNnxAN9+GrBs8zgABFEQIHgECnzbEA60wdJMBzGBSBobcwEwVAaBgJcaEIuUzFpZAAA4AAYYzIiB0gpcsaBzMjtywcpgJjABHBD4E5Wc2AAWg0HseyXfJ+NQfJmCHe8kEfMD0ioVAmxbVC2GguCBOApTQNQmBcHEKlTCopAxhjZtcFQPMGwcRSQG3ABWJcwQAFRM7MDJQmNylbABJKBu1gasCKImAMhC6scJg5D0mkaQgA)

```sh
--parser typescript
```

After switching from Prettier 2.2.1 to 2.3.0 in a large React + TS codebase, I've noticed an unexpectedly large diff. The offset for many React components has shifted.

It turns out that the length of the component name now affects the location of `({`. If these opening brackets are moved to the new line, the whole component body is shifted. This makes git diffs quite fragile.

Imagine renaming some component during a refactor. In Prettier 2.2.1, the diff would be just one line but in Prettier 2.3.0, `git blame` for the whole file would end up broken.

**Input:**

```tsx
interface MyComponentProps {
  x: number
}

const MyComponent: React.VoidFunctionComponent<MyComponentProps> = ({ x }) => {
  const a = useA()
  return <div>x = {x}; a = {a}</div>
}

interface MyComponent2Props {
  x: number
  y: number
}

const MyComponent2: React.VoidFunctionComponent<MyComponent2Props> = ({ x, y }) => {
  const a = useA()
  return <div>x = {x}; y = {y}; a = {a}</div>
}

interface MyComponentWithLongNameProps {
  x: number
  y: number
}

const MyComponentWithLongName: React.VoidFunctionComponent<MyComponentWithLongNameProps> = ({ x, y }) => {
  const a = useA()
  return <div>x = {x}; y = {y}; a = {a}</div>
}
```

**Output:**

```tsx
interface MyComponentProps {
  x: number;
}

const MyComponent: React.VoidFunctionComponent<MyComponentProps> = ({ x }) => {
  const a = useA();
  return (
    <div>
      x = {x}; a = {a}
    </div>
  );
};

interface MyComponent2Props {
  x: number;
  y: number;
}

const MyComponent2: React.VoidFunctionComponent<MyComponent2Props> = ({
  x,
  y,
}) => {
  const a = useA();
  return (
    <div>
      x = {x}; y = {y}; a = {a}
    </div>
  );
};

interface MyComponentWithLongNameProps {
  x: number;
  y: number;
}

const MyComponentWithLongName: React.VoidFunctionComponent<MyComponentWithLongNameProps> =
  ({ x, y }) => {
    const a = useA();
    return (
      <div>
        x = {x}; y = {y}; a = {a}
      </div>
    );
  };

```

**Expected behavior:**

All three `useA()` and `return` statements have only two leading spaces, i.e. the offset does not depend on the component name length.

Our team will be staying on Prettier 2.2.1 in the meantime. Happy to help with a PR if we come up with some solution!

### PR DESCRIPTION
**Pull request by thorn0**
## Description

This is an alternative to #10956. It uses the following formatting:
```ts
const MyComponent2: React.VoidFunctionComponent<
  MyComponent2Props
> = ({ x, y }) => {
  const a = useA();
  return (
    <div>
      x = {x}; y = {y}; a = {a}
    </div>
  );
};
```
@kachkaev How does this look to you?

Regression testing (https://github.com/prettier/prettier-regression-testing/issues/73#issuecomment-917767751) has shown that "prefer to keep the arrow signature on one line if possible" is a bad idea in this case.

Testing results for the current version of this PR: https://github.com/prettier/prettier-regression-testing/issues/73#issuecomment-918122605

Closes #10848

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/type-parameters.js [hunk 1]
```diff
@@ -32,12 +33,25 @@ function printTypeParameters(path, options, print, paramsKey) {
   const grandparent = path.getNode(2);
   const isParameterInTestCall = grandparent && isTestCall(grandparent);
 
+  const isArrowFunctionVariable = path.match(
```
Comment (thorn0) [inline src/language-js/print/type-parameters.js:36]: I don't like that there are two different `isArrowFunctionVariable` checks in two places. Looks like something that can get out of sync one day. Any idea how to improve this?

**Comment by kachkaev** [comment]
👋 @thorn0! The diff in integration tests makes absolute sense to me, so I’m more than happy to proceed with your solution 😍

File: src/language-js/print/type-parameters.js [hunk 2]
```diff
@@ -32,12 +33,25 @@ function printTypeParameters(path, options, print, paramsKey) {
...
+    (node) =>
+      !(node[paramsKey].length === 1 && isObjectType(node[paramsKey][0])),
+    undefined,
+    (node, name) => name === "typeAnnotation",
+    (node) => node.type === "Identifier",
+    (node) =>
+      node.type === "VariableDeclarator" &&
+      node.init &&
+      node.init.type === "ArrowFunctionExpression"
```
Comment (sosukesuzuki) [inline src/language-js/print/type-parameters.js:45]: Reuse `isArrowFunctionVariable`?

**Review by sosukesuzuki — COMMENTED** [review]
Overall looks good! But diffs are large, so I think we should release this as minor.

**Comment by kachkaev** [comment]
@sosukesuzuki this PR fixes a regression in Prettier 2.3.0 – the new behaviour closely matches 2.2.x. Given this and the fact that 2.4.0 is quite new, would it be reasonable to include the PR into 2.4.1?

If this was a completely new behaviour, I'd agree with you, but a former change between 2.2 and 2.3 tweaks the weights a bit I guess 🙂

**Review by alexander-akait — APPROVED** [review]
I like it

**Review by sosukesuzuki — APPROVED** [review]

**Comment by sosukesuzuki** [comment]
Hmmm... I still think this should be released in 2.5, as it is not a fix for the regressions that occurred in 2.4, and it would create a large diff. Many users will not like a big diff in a patched version. (But I'm not confident that this is the best way).

**Pull request merged**